src/mame/sinclair/atm.cpp

```diff
@@ -13,118 +23,379 @@ Not working because of banking issues.
 #include "specpls3.h"
 
 #include "beta_m.h"
-
+#include "bus/centronics/ctronics.h"
 #include "sound/ay8910.h"
 
-
 namespace {
 
+#define LOG_MEM   (1U << 1)
+#define LOG_VIDEO (1U << 2)
+#define LOG_WARN  (1U << 3)
+
+#define VERBOSE ( /*LOG_MEM | LOG_VIDEO |*/ LOG_WARN )
+#include "logmacro.h"
+
+#define LOGMEM(...)   LOGMASKED(LOG_MEM,   __VA_ARGS__)
+#define LOGVIDEO(...) LOGMASKED(LOG_VIDEO, __VA_ARGS__)
+#define LOGWARN(...)  LOGMASKED(LOG_WARN,  __VA_ARGS__)
+
+#define RAM_MASK 0x40
+#define DOS7FFD_MASK 0x80
+#define SHADOW m_beta->is_active()
+#define PEN_PAGE(bank) m_pages_map[BIT(m_port_7ffd_data, 4)][bank]
```

Comment:
Can you use a member function rather than a macro for this?  I don’t see anything problematic – just make it return a reference and you can use it in lvalue or rvalue context.